test/e2e/specs/editor/various/block-hierarchy-navigation.spec.js

```diff
@@ -0,0 +1,245 @@
+	test.beforeEach( async ( { admin } ) => {
+		await admin.createNewPost();
+	} );
+
+	test( 'should navigate using the list view sidebar', async ( {
+		editor,
+		page,
+		pageUtils,
+	} ) => {
+		await editor.openDocumentSettingsSidebar();
+		await editor.insertBlock( { name: 'core/columns' } );
+		await editor.canvas.click(
+			'role=button[name="Two columns; equal split"i]'
+		);
+
+		// Open the block inserter.
+		await page.keyboard.press( 'ArrowDown' );
+		await page.keyboard.press( 'Enter' );
+
+		// Add a paragraph in the first column.
+		const paragraph = page.getByRole( 'option', { name: 'Paragraph' } );
+		await expect( paragraph ).toBeVisible();
+		await paragraph.click();
+		await page.keyboard.type( 'First column' );
+
+		await pageUtils.pressKeys( 'access+o' );
+
+		const listView = page.getByRole( 'treegrid', {
+			name: 'Block navigation structure',
+		} );
+
+		await expect( listView ).toBeVisible();
+		await listView
+			.getByRole( 'gridcell', { name: 'Columns', exact: true } )
+			.click();
+
+		// Tweak the columns count.
+		await page.getByRole( 'spinbutton', { name: 'Columns' } ).click();
+		await pageUtils.pressKeys( 'primary+a' );
+		await page.keyboard.type( '3' );
+
+		// Wait for the new column block to appear in the list view
+		const column = listView.getByRole( 'gridcell', {
+			name: 'Column',
+			exact: true,
+		} );
+		await expect( column ).toHaveCount( 3 );
+
+		await column.last().click();
+
+		// Open the block inserter.
+		await page.keyboard.press( 'ArrowDown' );
+		await page.keyboard.press( 'Enter' );
+
+		await expect( paragraph ).toBeVisible();
+		await paragraph.click();
+		await page.keyboard.type( 'Third column' );
+
+		await expect.poll( editor.getBlocks ).toMatchObject( COLUMNS_BLOCK );
+	} );
+
+	test( 'should navigate block hierarchy using only the keyboard', async ( {
+		editor,
+		page,
+		pageUtils,
+	} ) => {
+		await editor.openDocumentSettingsSidebar();
+		await editor.insertBlock( { name: 'core/columns' } );
+		await editor.canvas.click(
+			'role=button[name="Two columns; equal split"i]'
+		);
+
+		// Open the block inserter.
+		await page.keyboard.press( 'ArrowDown' );
+		await page.keyboard.press( 'Enter' );
+
+		// Add a paragraph in the first column.
+		const paragraph = page.getByRole( 'option', { name: 'Paragraph' } );
+		await expect( paragraph ).toBeVisible();
+		await paragraph.click();
+		await page.keyboard.type( 'First column' );
+
+		await pageUtils.pressKeys( 'access+o' );
+		const listView = page.getByRole( 'treegrid', {
+			name: 'Block navigation structure',
+		} );
+		await expect( listView ).toBeVisible();
+
+		// Navigate to the columns blocks using the keyboard.
+		await pageUtils.pressKeys( 'ArrowUp', { times: 2 } );
+		await page.keyboard.press( 'Enter' );
+
+		// Move focus to the sidebar area.
+		await pageUtils.pressKeys( 'ctrl+`' );
+
+		// Navigate to the block settings sidebar and tweak the column count.
+		await pageUtils.pressKeys( 'Tab', { times: 5 } );
```

Comment:
The original test used `do...while` here, but I switched to a more straightforward approach. The test already relied on the "key * times" press in other areas.